From 4a1ee1bd32958150043ba530e0ebcb1cb46e159b Mon Sep 17 00:00:00 2001
From: squidbus <1249084-squidbus@users.noreply.gitlab.freedesktop.org>
Date: Wed, 29 Jul 2026 07:44:12 -0700
Subject: [PATCH] kk: Support VK_EXT_map_memory_placed

Using `mach_vm_remap`, we can create additional mappings from the address
provided by Metal, allowing us to support placed mappings.

Reviewed-by: Aitor Camacho <aitor@lunarg.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/43300>
---
 src/kosmickrisp/vulkan/kk_bo.c              | 60 +++++++++++++++++++++
 src/kosmickrisp/vulkan/kk_bo.h              |  5 ++
 src/kosmickrisp/vulkan/kk_device_memory.c   | 35 +++++++++---
 src/kosmickrisp/vulkan/kk_physical_device.c |  6 +++
 4 files changed, 100 insertions(+), 6 deletions(-)

diff --git a/src/kosmickrisp/vulkan/kk_bo.c b/src/kosmickrisp/vulkan/kk_bo.c
index 6926326969ed..6dea47500df9 100644
--- a/src/kosmickrisp/vulkan/kk_bo.c
+++ b/src/kosmickrisp/vulkan/kk_bo.c
@@ -12,6 +12,11 @@
 
 #include "util/u_memory.h"
 
+#if DETECT_OS_APPLE
+#include <mach/mach_init.h>
+#include <mach/mach_vm.h>
+#endif /* DETECT_OS_APPLE */
+
 VkResult
 kk_alloc_bo(struct kk_device *dev, struct vk_object_base *log_obj,
             uint64_t size_B, uint64_t align_B, struct kk_bo **bo_out)
@@ -88,3 +93,58 @@ kk_destroy_bo(struct kk_device *dev, struct kk_bo *bo)
 
    FREE(bo);
 }
+
+VkResult
+kk_bo_map_placed(struct kk_device *dev, struct kk_bo *bo, void **addr)
+{
+#if DETECT_OS_APPLE
+   mach_vm_address_t src_addr = (mach_vm_address_t)bo->cpu;
+   mach_vm_address_t dst_addr = (mach_vm_address_t)*addr;
+
+   int flags = VM_FLAGS_ANYWHERE;
+   if (dst_addr)
+      flags = VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE;
+
+   vm_prot_t cur_prot;
+   vm_prot_t max_prot;
+
+   /* Metal provides us with its own CPU mapping of the memory. mach_vm_remap
+    * allows us to create a second mapping pointing to that same memory, placed
+    * wherever the caller requests. Note that this is used even without a fixed
+    * address, since we need a unique mapping to support reserved unmap. */
+   kern_return_t ret = mach_vm_remap(mach_task_self(), &dst_addr, bo->size_B, 0,
+                                     flags, mach_task_self(), src_addr, false,
+                                     &cur_prot, &max_prot, VM_INHERIT_DEFAULT);
+   if (ret != KERN_SUCCESS)
+      return vk_errorf(dev, VK_ERROR_MEMORY_MAP_FAILED,
+                       "mach_vm_remap failed: %d", ret);
+
+   *addr = (void *)dst_addr;
+#endif /* DETECT_OS_APPLE */
+
+   return VK_SUCCESS;
+}
+
+VkResult
+kk_bo_unmap(struct kk_device *dev, struct kk_bo *bo, void *addr, bool reserved)
+{
+#if DETECT_OS_APPLE
+   mach_vm_address_t unmap_addr = (mach_vm_address_t)addr;
+   if (reserved) {
+      /* Replace with a reserved mapping using mach_vm_map */
+      kern_return_t ret = mach_vm_map(
+         mach_task_self(), &unmap_addr, bo->size_B, 0,
+         VM_FLAGS_FIXED | VM_FLAGS_OVERWRITE, MEMORY_OBJECT_NULL, 0, false,
+         VM_PROT_NONE, VM_PROT_READ | VM_PROT_WRITE, VM_INHERIT_DEFAULT);
+      if (ret != KERN_SUCCESS)
+         return vk_errorf(dev, VK_ERROR_MEMORY_MAP_FAILED,
+                          "mach_vm_map failed: %d", ret);
+   } else {
+      /* According to spec, we can only return a fail code here if the reserve
+       * flag is set, so ignore the result */
+      mach_vm_deallocate(mach_task_self(), unmap_addr, bo->size_B);
+   }
+#endif /* DETECT_OS_APPLE */
+
+   return VK_SUCCESS;
+}
diff --git a/src/kosmickrisp/vulkan/kk_bo.h b/src/kosmickrisp/vulkan/kk_bo.h
index 6ea5ca6693da..3280eb614da8 100644
--- a/src/kosmickrisp/vulkan/kk_bo.h
+++ b/src/kosmickrisp/vulkan/kk_bo.h
@@ -12,6 +12,7 @@
 #include "vulkan/vulkan_core.h"
 
 #include <inttypes.h>
+#include <stdbool.h>
 
 struct kk_device;
 struct vk_object_base;
@@ -38,4 +39,8 @@ VkResult kk_alloc_bo(struct kk_device *dev, struct vk_object_base *log_obj,
 
 void kk_destroy_bo(struct kk_device *dev, struct kk_bo *bo);
 
+VkResult kk_bo_map_placed(struct kk_device *dev, struct kk_bo *bo, void **addr);
+VkResult kk_bo_unmap(struct kk_device *dev, struct kk_bo *bo, void *addr,
+                     bool reserved);
+
 #endif /* KK_BO_H */
diff --git a/src/kosmickrisp/vulkan/kk_device_memory.c b/src/kosmickrisp/vulkan/kk_device_memory.c
index 0e11cc2479bc..c70a8e7491fe 100644
--- a/src/kosmickrisp/vulkan/kk_device_memory.c
+++ b/src/kosmickrisp/vulkan/kk_device_memory.c
@@ -166,6 +166,10 @@ kk_FreeMemory(VkDevice device, VkDeviceMemory _mem,
    if (!mem)
       return;
 
+   /* According to spec, memory is implicitly unmapped upon free */
+   if (mem->map)
+      kk_bo_unmap(dev, mem->bo, mem->map, false);
+
    kk_destroy_bo(dev, mem->bo);
 
    vk_device_memory_destroy(&dev->vk, pAllocator, &mem->vk);
@@ -188,6 +192,13 @@ kk_MapMemory2KHR(VkDevice device, const VkMemoryMapInfoKHR *pMemoryMapInfo,
    const VkDeviceSize size = vk_device_memory_range(
       &mem->vk, pMemoryMapInfo->offset, pMemoryMapInfo->size);
 
+   void *fixed_addr = NULL;
+   if (pMemoryMapInfo->flags & VK_MEMORY_MAP_PLACED_BIT_EXT) {
+      const VkMemoryMapPlacedInfoEXT *placed_info = vk_find_struct_const(
+         pMemoryMapInfo->pNext, MEMORY_MAP_PLACED_INFO_EXT);
+      fixed_addr = placed_info->pPlacedAddress;
+   }
+
    /* From the Vulkan spec version 1.0.32 docs for MapMemory:
     *
     *  * If size is not equal to VK_WHOLE_SIZE, size must be greater than 0
@@ -213,9 +224,13 @@ kk_MapMemory2KHR(VkDevice device, const VkMemoryMapInfoKHR *pMemoryMapInfo,
                        "Memory object already mapped.");
    }
 
-   // TODO_KOSMICKRISP Use mmap here to so we can support VK_EXT_map_memory_placed
-   mem->map = mem->bo->cpu;
+   void *mapped_addr = fixed_addr;
+   result = kk_bo_map_placed(dev, mem->bo, &mapped_addr);
+   if (result != VK_SUCCESS)
+      return result;
 
+   assert(!fixed_addr || mapped_addr == fixed_addr);
+   mem->map = mapped_addr;
    *ppData = mem->map + offset;
 
    return result;
@@ -225,16 +240,24 @@ VKAPI_ATTR VkResult VKAPI_CALL
 kk_UnmapMemory2KHR(VkDevice device,
                    const VkMemoryUnmapInfoKHR *pMemoryUnmapInfo)
 {
+   VK_FROM_HANDLE(kk_device, dev, device);
    VK_FROM_HANDLE(kk_device_memory, mem, pMemoryUnmapInfo->memory);
+   VkResult result = VK_SUCCESS;
 
    if (mem == NULL)
-      return VK_SUCCESS;
+      return result;
+
+   bool reserved = pMemoryUnmapInfo->flags & VK_MEMORY_UNMAP_RESERVE_BIT_EXT;
+
+   result = kk_bo_unmap(dev, mem->bo, mem->map, reserved);
+   if (result != VK_SUCCESS)
+      return result;
 
-   // TODO_KOSMICKRISP Use unmap here to so we can support
-   // VK_EXT_map_memory_placed
+   /* According to spec, memory is not unmapped on failure, so we only clear to
+    * NULL on success */
    mem->map = NULL;
 
-   return VK_SUCCESS;
+   return result;
 }
 
 VKAPI_ATTR VkResult VKAPI_CALL
diff --git a/src/kosmickrisp/vulkan/kk_physical_device.c b/src/kosmickrisp/vulkan/kk_physical_device.c
index 412af3fc53f5..d7020a8cfa8e 100644
--- a/src/kosmickrisp/vulkan/kk_physical_device.c
+++ b/src/kosmickrisp/vulkan/kk_physical_device.c
@@ -180,6 +180,7 @@ kk_get_device_extensions(const struct kk_instance *instance,
       .EXT_hdr_metadata = true,
       .EXT_image_2d_view_of_3d = true,
       .EXT_load_store_op_none = true,
+      .EXT_map_memory_placed = true,
       .EXT_memory_budget = true,
       .EXT_multi_draw = true,
       .EXT_mutable_descriptor_type = true,
@@ -458,6 +459,11 @@ kk_get_device_features(
       .image2DViewOf3D = true,
       .sampler2DViewOf3D = true,
 
+      /* VK_EXT_map_memory_placed */
+      .memoryMapPlaced = true,
+      .memoryMapRangePlaced = false,
+      .memoryUnmapReserve = true,
+
       /* VK_EXT_multi_draw */
       .multiDraw = true,
 
-- 
GitLab

